agent/command/papertrail_trace.go

```diff
@@ -0,0 +1,213 @@
+	comm agentinternalclient.Communicator, logger agentinternalclient.LoggerProducer, conf *agentinternal.TaskConfig) error {
+	if err := evergreenutil.ExpandValues(t, &conf.Expansions); err != nil {
+		return fmt.Errorf("apply expansions: %w", err)
+	}
+
+	client := &papertrailClient{
+		httpc:     &http.Client{},
+		address:   t.Address,
+		keyID:     t.KeyID,
+		secretKey: t.SecretKey,
+	}
+
+	task := conf.Task
+
+	papertrailBuildID := fmt.Sprintf("%s_%d", task.Id, task.Execution)
+
+	workdir := t.WorkDir
+	if workdir == "" {
+		workdir = conf.WorkDir
+	}
+
+	for _, file := range t.Filenames {
+		fullname := filepath.Join(workdir, file)
+
+		sha256sum, err := getSHA256(fullname)
+		if err != nil {
+			return fmt.Errorf("get sha256: %w", err)
+		}
+
+		// should already be the basename, but just in case
+		basename := filepath.Base(file)
+
+		args := traceArgs{
+			build:     papertrailBuildID,
+			platform:  "evergreen",
+			filename:  basename,
+			sha256:    sha256sum,
+			product:   t.Product,
+			version:   t.Version,
+			submitter: task.ActivatedBy,
+		}
+
+		if err := client.trace(ctx, args); err != nil {
+			return fmt.Errorf("run trace: %w", err)
+		}
+
+		logger.Task().Infof("Successfully traced '%s' (sha256://%s)", fullname, sha256sum)
+	}
+
+	return nil
+}
+
+func getSHA256(filename string) (string, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return "", fmt.Errorf("open file: %w", err)
+	}
+
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", fmt.Errorf("read file: %w", err)
+	}
+
+	sha256sum := fmt.Sprintf("%x", h.Sum(nil))
+
+	return sha256sum, nil
+}
+
+func (t *papertrailTrace) ParseParams(params map[string]any) error {
+	conf := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           t,
+	}
+
+	decoder, err := mapstructure.NewDecoder(conf)
+	if err != nil {
+		return fmt.Errorf("new decoder: %w", err)
+	}
+
+	if err := decoder.Decode(params); err != nil {
+		return fmt.Errorf("decode params: %w", err)
+	}
+
+	if t.Address == "" {
+		t.Address = productionPapertrailAddress
+	}
+
+	if t.KeyID == "" {
+		return errMustSpecifyKeyID
+	}
+
+	if t.SecretKey == "" {
+		return errMustSpecifySecretKey
+	}
+
+	if t.Product == "" {
+		return errMustSpecifyProduct
+	}
+
+	if t.Version == "" {
+		return errMustSpecifyVersion
+	}
+
+	if len(t.Filenames) == 0 {
+		return errMustSpecifyAtLeastOneFilename
+	}
+
+	return nil
+}
+func (t *papertrailTrace) Name() string { return "papertrail.trace" }
+
+func papertrailTraceFactory() Command { return &papertrailTrace{} }
+
+type papertrailClient struct {
+	httpc     *http.Client
+	address   string
+	keyID     string
+	secretKey string
+}
+
+type traceArgs struct {
+	build     string
+	platform  string
+	filename  string
+	sha256    string
+	product   string
+	version   string
+	submitter string
+}
+
+func (c *papertrailClient) trace(ctx context.Context, args traceArgs) error {
```

Comment:
Sorry, I'm not sure what you mean that this would be better done on the agent rather than on the client. Are you saying that the agent shouldn't be doing this work?

Reply:
Yeah essentially, although we could still. There are two reasons I can think for it to be better for the thirdparty function to be called in a rest route as opposed to directly in the agent:
1) Better logging. Do you want errors to go to Splunk or just the task logs? If the latter, then the agent is fine.
2) Better security -- I see right now you have the key/secret as user inputted, but to verify, do you expect this to be different for users, or will you be configuring the same value for all release-relevant projects? In which case we should just define this globally on the admin page and get those values in the rest route; this way the agent never needs access to the secrets.

Reply:
1. Errors should go to only the task logs and nowhere else
2. The key ID and secret key will be different for each project

Reply:
Okay calling from the agent seems okay then, with moving the logic into the thirdparty package

Reply:
Moved the client stuff into the thirdparty package!